### DISCUSSION TITLE
Uno.Extension-Navigation: use existing ViewModel

### DISCUSSION THREAD

**Discussion by TopperDEL** [unoplatform/uno#8867] [Q&A] [answered]
How is the following with Uno.Extensions supposed to work: if I have a ListView bound to a collection of "ItemViewModels" and click one entry - how can I navigate to a View that gets the DataContext set to that selected ItemViewModel (instead of creating a new/empty one for the view)?

I currently send the selected ViewModel as a parameter to the view and set the DataContext manually. But that feels not right.

**Comment by nickrandolph** [comment, marked as answer]
Hey, check out this how-to guide and let me know if you still need a hand: https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Tutorials/Navigation/HowTo-DisplayItem.html

**Reply by TopperDEL** [reply]
Thanks! But that does not completely answer my question. My ListView is not bound to a "model-object", but already bound to a "view-model" that I want to use as the base (DataContext) for the view I navigate to. Maybe I'm turning in circles. But I just want to do like ``Navigator.NavigateViewModel(this, myAlreadyCreatedViewModel)`` and have that one assigned as DataContext of the View mapped to "myAlreadyCreatedViewModel".

**Reply by nickrandolph** [reply]
Currently we don't support binding to an existing viewmodel in this way (see issue logged on extensions repo: https://github.com/unoplatform/uno.extensions/issues/496). My suggestion in the short term is to create a viewmodel for the destination view that accepts you ItemViewModel as an parameter in the constructor

**Reply by francoistanguay** [reply]
@TopperDEL  Can you share your scenario where you want to reuse ViewModel types (and instances) between ListView Item and "Details" page?

What we've seen typically is that ViewModels are slightly different where the Item represents a Summary and then you navigate to a page for more details. So always curious to see scenarios where it makes sense.

**Reply by TopperDEL** [reply]
For me it's quite normal. Imagine a list of e.g. "students". I might have the the ListView bound to Students-DTO's directly, which will be ok in most cases. But if I need some conversion on the Student-object to show some special data in the list view, I would wrap it in a StudentsViewModel. But if I then want to get into the Students Details by clicking a student in the list, it might use the same StudentViewModel but just display additional data from that ViewModel.

I might have a "StudentListViewModel" and a "StudentDetailsViewModel" in this case - but that feels over-engineered. What would be your path to got here? Maybe I'm doing it "wrong"?

**Reply by ToddGlodek** [reply]
Maybe something like this will work for you 

```
        private void NavigateToOtherPage()
        {
            try
            {
                NavContent.Navigate(typeof(OtherPage));
                var viewModel = ((NavContent.Content as OtherPage).DataContext as OtherPageViewModel);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }
```

**Reply by TopperDEL** [reply]
Yeah, I'm doing the opposit quite now - sending my existing ViewModel as nav-parameter and setting it in the View OnNavigatedTo. The above one is better, though, as the OtherPage here does not need to be aware of the OtherPagewViewModel.

**Reply by nickrandolph** [reply]
@TopperDEL The PR which allows for using an existing ViewModel instance has been merged (see https://github.com/unoplatform/uno.extensions/issues/496). If you update to the latest dev packages for uno extensions (update all packages with prefix Uno.Extensions) it would be great to know whether this addresses your issue?

**Reply by TopperDEL** [reply]
Thanks @nickrandolph! But unfortunately that does not yet seem to help in my case. But maybe I'm doing something wrong? I'm doing the navigation in my ShellViewModel, as I'm using a MessageBus here. So the view itself does not navigate. I do it like this:

``
await Navigator.NavigateViewAsync<RoomView>(this, data: eventData.RoomViewModel);
``

I have a RouteMap like this in place:
```
new ViewMap<RoomView, RoomViewModel>(),
[...]
new RouteMap("RoomView", View: views.FindByViewModel<RoomViewModel>())
```

I get to the correct view, but the DataContext hasthe ShellViewModel set (in OnNavigatedTo) which I think is kind of like a default here.

**Reply by nickrandolph** [reply]
yeh so because of the way WinUI (UWP) does frame navigation the viewmodel isn't necessarily set as the datacontext at the time the onnavigatedto is called. If you want to access the viewmodel in the codebehind, you need to attach an event handler to the DataContextChanged event.

The code you have should be correct - can you put a breakpoint in the RoomViewModel constructor and make sure it doesn't get called during navigation?

**Reply by TopperDEL** [reply]
Oh, indeed - it is working. I've mixed it up together with my old code to handle the Navigation-data. Awesome! now it is like it should be. Thank you!
